File: solutions/0001-two-sum/two-pointers.py
#
# @title: Two Pointers
# @time: O(n log n)
# @space: O(n)
#

def twoSum(nums, target):
    """
    nums: list[int]
    target: int
    return: list[int]
    """
    nums = [[i, num] for i, num in enumerate(nums)]
    nums.sort(key=lambda x: x[1])

    left, right = 0, len(nums) - 1

    while left < right:
        sum_ = nums[left][1] + nums[right][1]
        if sum_ == target:
            return [nums[left][0], nums[right][0]]
        if sum_ > target:
            right -= 1
        else:
            left += 1

File: solutions/0001-two-sum/two-pointers.js
// time: O(n log n)
// space: O(n)

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSum = (nums, target) => {
    nums = nums.map((num, id) => [id, num]);
    nums.sort((a, b) => a[1] - b[1]);

    let left = 0,
        right = nums.length - 1;

    while (left < right) {
        const sum = nums[left][1] + nums[right][1];
        if (sum === target) return [nums[left][0], nums[right][0]];
        sum > target ? right-- : left++;
    }
};
